### PR TITLE
Support diplomat::attr(auto)

### PR DESCRIPTION
**Pull request by Manishearth**
Builds on top of https://github.com/rust-diplomat/diplomat/pull/584

This is the change I proposed in #563

This makes it clearer as to when the attribute is automatically deciding which backend to target, and brings back some error messages when you try and unconditionally force an unsupported attribute on a backend.

### CODE REVIEW

File: feature_tests/src/attrs.rs [hunk 1]
```diff
@@ -88,25 +88,25 @@ pub mod ffi {
     pub struct MyIndexer(Vec<String>);
```
Comment (robertbastian) [inline feature_tests/src/attrs.rs:88]: should also be `auto`, as all other `any(_)`s

Comment (Manishearth) [reply feature_tests/src/attrs.rs:88]: I don't think `auto` works that way

Comment (robertbastian) [reply feature_tests/src/attrs.rs:88]: I thought that's the point?

Comment (Manishearth) [reply feature_tests/src/attrs.rs:88]: However I should use real `not(any(supports ...))`s here instead

Comment (robertbastian) [reply feature_tests/src/attrs.rs:87]: ah yeah this is disable

Comment (Manishearth) [reply feature_tests/src/attrs.rs:87]: @robertbastian No, it's only "auto" based on the actual attribute being used. It won't affect `disable`

Comment (Manishearth) [reply feature_tests/src/attrs.rs:87]: I plan to document this in the diplomat book later when I'm back, but the idea is that `auto` saves you from doing redundant `attr(supports = foo, foo)`. If you're doing something more complicated, like "this type only exists for backends with iterators", you still need to do `not(supports = iterators), disable` etc. Which was why I resurrected these supports attributes in https://github.com/rust-diplomat/diplomat/pull/584

Comment (robertbastian) [reply feature_tests/src/attrs.rs:87]: there are still explicit backends in this file, please recover them as well. Might want to do `git reset` to f0f1a1a746fa87020bb7b7397d92ba0c60db6fc7 and work from there.

Comment (Manishearth) [reply feature_tests/src/attrs.rs:88]: Looks like this test is currently disabled in all backends anyway. I'm going to leave it in for when JS gets indexer support, but it appears to be that Dart indexer support is broken around indexers that return strings.

Comment (robertbastian) [reply feature_tests/src/attrs.rs:88]: How so?

Comment (Manishearth) [reply feature_tests/src/attrs.rs:88]: https://github.com/rust-diplomat/diplomat/issues/590

Comment (Manishearth) [reply feature_tests/src/attrs.rs:88]: It's unrelated to indexers, it's about returning `&DiplomatStr`.

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/MyIndexer.kt
[the remark this answers is not in the document]
Comment (Manishearth) [reply feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/MyIndexer.kt:1]: yeah it got deleted in a previous commit and I didn't git add it again. It's back

**Comment by Manishearth** [comment]
I think this covers all explicit backends except for the `not(c)` at the top, and C is special anyway

**Comment by Manishearth** [comment]
We can continue to clean up the tests later too. Ideally all explicit backends in tests are for these reasons only:

 - Bugs in specific backends, like https://github.com/rust-diplomat/diplomat/issues/590 (I'm so glad `disable` works seamlessly so that now I can work on individual backends without having to get distracted by finding bugs like that!)
 - Differing conventions. We _may_ wish to introduce a separate gate that lets us talk about groups of conventions, like the C++ namespace-everything convention vs the JS/Dart your-package-is-already-namespaced convention
 - C backend specific shenanigans.

**Comment by robertbastian** [comment]
> Differing conventions. We may wish to introduce a separate gate that lets us talk about groups of conventions, like the C++ namespace-everything convention vs the JS/Dart your-package-is-already-namespaced convention

In the end I believe that languages are different enough that modelling their behaviour by simple `support = ` bools is a lost cause, and for best results you should do backend-specific things. For example, Dart supports getters and setters, but it requires that for the same field name, they're the same type. We could add a cfg for that, but that's so niche that it should just be `cfg(dart)`. We only have 4 backends in ICU4X, it's fine to do custom things.

**Comment by Manishearth** [comment]
(to be clear I was talking about our tests there)

> We only have 4 backends in ICU4X, it's fine to do custom things.

Right, but I want to make Diplomat flexible for other users. Also ICU4X has definitely expressed interest in a JVM and a Python backend.

I think in ICU4X we can and should be nitpicky and specific about `cfg(backend)` but the `supports` gates are useful where necessary (especially around iterators, which are messy to do manually!), and we could even consider introducing `conventions = foobar` gates for any cases we start doing `any(backend1, backend2)` (I don't wish to overload `supports = ` for things that are conventions).

I definitely don't consider "use `supports` everywhere" to be a rule I wish to apply to ICU4X. Where it's straightforward, yes, because it's cleaner, and `auto` will help us here too, but we don't need to try too hard to avoid explicit backends. I'd say for ICU4X the main goal should be making the attributes not too confusing, I don't want `attr(any(js, dart), ...)` to be something that is non-obvious to the reader.

**Pull request merged**